fusestore/store/src/meta_service/state_machine.rs

```diff
@@ -212,25 +220,85 @@ impl StateMachine {
                 }
             }
 
-            Cmd::AddDatabase { ref name } => {
+            Cmd::CreateDatabase {
+                ref name,
+                ref if_not_exists,
+                ..
+            } => {
                 // - If the db present, return it.
                 // - Otherwise, create a new one with next seq number as database id, and add it in to store.
                 if self.databases.contains_key(name) {
                     let prev = self.databases.get(name);
-                    Ok((prev.cloned(), None).into())
+                    if *if_not_exists {
+                        Ok((prev.cloned(), None).into())
+                    } else {
+                        Ok((None::<Database>, None::<Database>).into())
+                    }
                 } else {
                     let db = Database {
                         database_id: self.incr_seq(SEQ_DATABASE_ID),
                         tables: Default::default(),
                     };
 
                     let prev = self.databases.insert(name.clone(), db.clone());
-                    tracing::debug!("applied AddDatabase: {}={:?}", name, db);
+                    tracing::debug!("applied CreateDatabase: {}={:?}", name, db);
 
                     Ok((prev, Some(db)).into())
                 }
             }
 
+            Cmd::DropDatabase { ref name } => {
+                let prev = self.databases.get(name).cloned();
+                self.databases.remove(name);
+                tracing::debug!("applied DropDatabase: {}", name);
+                Ok((prev, None).into())
+            }
+
+            Cmd::CreateTable {
+                ref db_name,
+                ref table_name,
+                ref if_not_exists,
+                ref table,
+            } => {
+                let db = self.databases.get(db_name);
+                let mut db = db.unwrap().to_owned();
+
+                if db.tables.contains_key(table_name) {
+                    let table_id = db.tables.get(table_name).unwrap();
+                    let prev = self.tables.get(table_id);
+                    if *if_not_exists {
+                        Ok((prev.cloned(), None).into())
+                    } else {
+                        Ok((None::<Table>, None::<Table>).into())
+                    }
+                } else {
```

Comment:
The returned value should reflect the state change, i.e., includes the state before applying the cmd and the state after applying the cmd.

Thus if nothing is done, it should returns `(prev, prev)` to indicate that nothing changed.

`(prev, None)` could be a state change after applying a delete operation.

Reply:
> The returned value should reflect the state change, i.e., includes the state before applying the cmd and the state after applying the cmd.
> 
> Thus if nothing is done, it should returns `(prev, prev)` to indicate that nothing changed.
> 
> `(prev, None)` could be a state change after applying a delete operation.

Sorry for the above statement. I just found some old codes did not follow this: `(prev, None)` was used to indicate a failed `add` operation. Thus the caller can not tell the difference between a failed `add` and a successful `delete`.

An solution would be add a `seq` to every value(for now `seq` is only used by `Cmd::UpsertKV`) and `apply()` returns the state before and after applying a `Cmd`.
Thus a failed `add` returns `((seq_1, value), (seq_1, value))` while a successful `delete` returns `((seq_1, value), None)`.

Any idea? 🤔 @ariesdevil

Reply:
Or we can use something like [this one](https://lib.rs/crates/enum_index) in return values to indicate the different commands.

Also, add a `seq` is fine for me.

Reply:
Do these changes in this area will affect changes to the API, should I leave a todo marker here and create a new issue for this? @drmingdrmer

Reply:
> Do these changes in this area will affect changes to the API, should I leave a todo marker here and create a new issue for this? @drmingdrmer

Absolutely yes. Another issue seems a great idea!

Reply:
> Or we can use something like [this one](https://lib.rs/crates/enum_index) in return values to indicate the different commands.
> 
> Also, add a `seq` is fine for me.

This crate is interesting and what I've been looking for!
Seems another issue is coming 🤔 ?